securedrop/tests/test_i18n.py

```diff
@@ -0,0 +1,148 @@
+import i18n
+import journalist
+import manage
+import source
+import version
+
+
+class TestI18N(object):
+
+    def verify_i18n(self, app):
+        not_translated = 'code hello i18n'
+        translated_fr = 'code bonjour'
+
+        for accepted in ('unknown', 'en_US'):
+            headers = Headers([('Accept-Language', accepted)])
+            with app.test_request_context(headers=headers):
+                assert not hasattr(request, 'babel_locale')
+                assert not_translated == gettext(not_translated)
+                assert hasattr(request, 'babel_locale')
+                assert render_template_string('''
+                {{ gettext('code hello i18n') }}
+                ''').strip() == not_translated
+
+        for lang in ('fr_FR', 'fr', 'fr-FR'):
+            headers = Headers([('Accept-Language', lang)])
+            with app.test_request_context(headers=headers):
+                assert not hasattr(request, 'babel_locale')
+                assert translated_fr == gettext(not_translated)
+                assert hasattr(request, 'babel_locale')
+                assert render_template_string('''
+                {{ gettext('code hello i18n') }}
+                ''').strip() == translated_fr
+
+        translated_cn = 'code chinese'
+
+        for lang in ('zh-CN', 'zh-Hans-CN'):
+            headers = Headers([('Accept-Language', lang)])
+            with app.test_request_context(headers=headers):
+                assert not hasattr(request, 'babel_locale')
+                assert translated_cn == gettext(not_translated)
+                assert hasattr(request, 'babel_locale')
+                assert render_template_string('''
+                {{ gettext('code hello i18n') }}
+                ''').strip() == translated_cn
+
+        with app.test_client() as c:
+            c.get('/')
+            assert session.get('locale') is None
+            assert not_translated == gettext(not_translated)
+
+            c.get('/?l=fr_FR', headers=Headers([('Accept-Language', 'en_US')]))
+            assert session.get('locale') == 'fr_FR'
+            assert translated_fr == gettext(not_translated)
+
+            c.get('/', headers=Headers([('Accept-Language', 'en_US')]))
+            assert session.get('locale') == 'fr_FR'
+            assert translated_fr == gettext(not_translated)
+
+            c.get('/?l=')
+            assert session.get('locale') is None
+            assert not_translated == gettext(not_translated)
+
+            c.get('/?l=en_US', headers=Headers([('Accept-Language', 'fr_FR')]))
+            assert session.get('locale') == 'en_US'
+            assert not_translated == gettext(not_translated)
+
+            c.get('/', headers=Headers([('Accept-Language', 'fr_FR')]))
+            assert session.get('locale') == 'en_US'
+            assert not_translated == gettext(not_translated)
+
+            c.get('/?l=', headers=Headers([('Accept-Language', 'fr_FR')]))
+            assert session.get('locale') is None
```

Comment:
Why would we want `locale` to be `None` and not `en_US` here?

Reply:
Because the session is not always set. It is set when **l=XXX** and unset when **l=**